speculation-rules.bs

```diff
@@ -300,6 +427,40 @@ A <dfn>prerender candidate</dfn> is a [=struct=] with the following [=struct/ite
   We should also cancel speculated prerenders.
 </p>
 
+<h3 id="document-rule-predicate-matching">Document rule predicate matching</h3>
+
+<div algorithm>
+  A [=document rule predicate=] |predicate| <dfn for="document rule predicate">matches</dfn> an {{Element}} |el| implementing the {{HTMLHyperlinkElementUtils}} mixin if the following steps return true:
+
+  1. If |predicate| is a [=document rule conjunction=], then:
+    1. [=list/For each=] |clause| of |predicate|'s [=document rule conjunction/clauses=]:
+      1. If |clause| does not [=document rule predicate/match=] |el|, return false.
+    1. Return true.
+  1. If |predicate| is a [=document rule disjunction=], then:
+    1. [=list/For each=] |clause| of |predicate|'s [=document rule disjunction/clauses=]:
+      1. If |clause| [=document rule predicate/matches=] |el|, return true.
+    1. Return false.
+  1. If |predicate| is a [=document rule negation=], then:
+    1. If |predicate|'s [=document rule negation/clause=] [=document rule predicate/matches=] |el|, return false.
+    1. Return true.
+  1. If |predicate| is a [=document rule URL pattern predicate=], then:
+    1. Let |href| be the result of running |el|'s {{HTMLHyperlinkElementUtils/href}} getter steps.
+    1. [=list/For each=] |pattern| of |predicate|'s [=document rule URL pattern predicate/patterns=]:
+      1. Execute |pattern|'s {{URLPattern/test(input, baseURL)}} steps given |href|. If the reuslt is true, return true.
+    1. Return false.
+  1. If |predicate| is a [=document rule CSS selector predicate=], then:
+    1. [=list/For each=] |selector| of |predicate|'s [=document rule CSS selector predicate/selectors=]:
+      1. [=match a selector against an element|Match=] |selector| against |el| with [=scoping root=] |el|'s [=Node/node document=]. If the result is true, return true.
```

Comment:
```suggestion
      1. [=match a selector against an element|Match=] |selector| against |el| with the [=scoping root=] set to |el|'s [=Node/node document=]. If the result is true, return true.
```

Although, I wonder if it should be not supplied? Or set to el's root? Not supplied means you can't use `:scope` I think, which is perhaps simplest?

Reply:
Or alternatively just making this match `Element.matches()`, i.e. set scope to |el|?

Reply:
I admit to not seeing `:scope` being used much and therefore not having a good intuition for it.

My gut is that what we're doing here is fundamentally more similar to `document.querySelectorAll` (i.e., we're searching for matching elements within the document when we consider speculation). That algorithm uses [scope-match a selectors string](https://dom.spec.whatwg.org/#scope-match-a-selectors-string) which provides the document itself as the scoping root. (Aside: I can't find in the spec why the root element matches `:scope` despite the document being provided as the scoping root, but this does seem to be the case.)

`Element.matches` would be the obvious counterexample, but I think that's different, because as far as the API user is concerned, they're searching the document, not particular elements. I could be convinced otherwise, though.

So it seems to me the motivating factor should be "what, to the author, is the scope within which their selector applies"? I think that's the document whose tree we're walking.